provider/pkg/provider/remote/connection.go

```diff
@@ -122,8 +128,51 @@ func (con *Connection) Dial(ctx p.Context, config *ssh.ClientConfig) (*ssh.Clien
 	_, _, err = retry.Until(ctx, retry.Acceptor{
 		Accept: func(try int, nextRetryTime time.Duration) (bool, interface{}, error) {
 			client, err = ssh.Dial("tcp",
-				net.JoinHostPort(*con.Host, fmt.Sprintf("%.0f", *con.Port)),
+				net.JoinHostPort(*con.Host, fmt.Sprintf("%d", int(*con.Port))),
 				config)
+			if con.Proxy != nil && *con.Proxy.Host != "" {
+				proxyConfig, err := con.Proxy.SShConfig()
+				if err != nil {
+					return false, nil, err
+				}
+				proxyClient, err := ssh.Dial("tcp",
+					net.JoinHostPort(*con.Proxy.Host, fmt.Sprintf("%d", int(*con.Proxy.Port))),
+					proxyConfig)
+				if err != nil {
+					// on each try we already made a dial
+					dials := try + 1
+					if reachedDialingErrorLimit(dials, dialErrorLimit) {
+						return true, nil, err
+					}
+					return false, nil, nil
+				}
+
+				endpoint := net.JoinHostPort(*con.Host, fmt.Sprintf("%d", int(*con.Port)))
+				netConn, err := proxyClient.Dial("tcp", endpoint)
+				if err != nil {
+					// on each try we already made a dial
+					dials := try + 1
```

Comment:
Is it intentional that we aren't applying the same logging and timeout adjustments on these paths?  I don't quite follow why those are necessary on the other branch - but seems we may want to be consistent between the two?

Reply:
We should. I'll separate the retry logic out into another function, then apply it as appropriate to all connections.